DevTools/Playground/main.cpp

```diff
@@ -25,12 +25,139 @@
+public:
+    GMLAutocompleteProvider() { }
+    virtual ~GMLAutocompleteProvider() override { }
+
+private:
+    virtual void provide_completions(Function<void(Vector<Entry>)> callback) override
+    {
+        auto cursor = m_editor->cursor();
+        auto text = m_editor->text();
+        GUI::GMLLexer lexer(text);
+        // FIXME: Provide a begin() and end() for lexers PLEASE!
+        auto all_tokens = lexer.lex();
+        enum State {
+            Free,
+            InClassName,
+            AfterClassName,
+            InIdentifier,
+            AfterIdentifier, // Can we introspect this?
+        } state { Free };
+        String identifier_string;
+        Vector<String> class_names;
+        Vector<State> previous_states;
+        bool should_push_state { true };
+
+        for (auto& token : all_tokens) {
+            if (token.m_start.line > cursor.line() || (token.m_start.line == cursor.line() && token.m_start.column > cursor.column()))
+                break;
+
+            switch (state) {
+            case Free:
+                if (token.m_type == GUI::GMLToken::Type::ClassName) {
+                    if (should_push_state)
+                        previous_states.append(state);
+                    else
+                        should_push_state = true;
+                    state = InClassName;
+                    class_names.append(token.m_view);
+                    break;
+                }
+                break;
+            case InClassName:
+                state = AfterClassName;
+                break;
+            case AfterClassName:
+                if (token.m_type == GUI::GMLToken::Type::Identifier) {
+                    state = InIdentifier;
+                    identifier_string = token.m_view;
+                    break;
+                }
+                if (token.m_type == GUI::GMLToken::Type::RightCurly) {
+                    class_names.take_last();
+                    state = previous_states.take_last();
+                    break;
+                }
+                if (token.m_type == GUI::GMLToken::Type::ClassMarker) {
+                    previous_states.append(AfterClassName);
+                    state = Free;
+                    should_push_state = false;
+                }
+                break;
+            case InIdentifier:
+                state = AfterIdentifier;
+                break;
+            case AfterIdentifier:
+                if (token.m_type == GUI::GMLToken::Type::ClassMarker) {
+                    previous_states.append(AfterClassName);
+                    state = Free;
+                    should_push_state = false;
+                } else {
+                    state = AfterClassName;
+                }
+                break;
+            }
+        }
+
+        Vector<GUI::AutocompleteProvider::Entry> entries;
+        dbg() << "In state " << (int)state << ", have " << class_names.size() << " class names, last seen identifier " << identifier_string;
```

Comment:
Please use `dbgln()` in new code

Reply:
I've removed that, it's not necessary - just a leftover.